src/util/ALE.sol

```diff
@@ -0,0 +1,459 @@
+
+        if (_buySellToken != _collateral) {
+            IERC20(_collateral).approve(_market, type(uint256).max);
+        }
+
+        if (_helper != address(0)) {
+            markets[_market].helper = ITransformHelper(_helper);
+            IERC20(_buySellToken).approve(_helper, type(uint256).max);
+            IERC20(_collateral).approve(_helper, type(uint256).max);
+        }
+
+        emit NewMarket(_market, _buySellToken, _collateral, _helper);
+    }
+
+    /// @notice Update the helper contract
+    /// @param _market The market we want to update the helper contract for
+    /// @param _helper The helper contract
+    function updateMarketHelper(
+        address _market,
+        address _helper
+    ) external onlyOwner {
+        address oldHelper = address(markets[_market].helper);
+        if (oldHelper != address(0)) {
+            markets[_market].buySellToken.approve(oldHelper, 0);
+            markets[_market].collateral.approve(oldHelper, 0);
+        }
+
+        markets[_market].helper = ITransformHelper(_helper);
+        markets[_market].buySellToken.approve(_helper, type(uint256).max);
+        markets[_market].collateral.approve(_helper, type(uint256).max);
+
+        emit NewHelper(_market, _helper);
+    }
+
+    /// @notice Leverage user position by minting DOLA, buying collateral, deposting into the user escrow and borrow DOLA on behalf to repay the minted DOLA
+    /// @dev Requires user to sign message to permit the contract to borrow DOLA on behalf
+    /// @param _value Amount of DOLA to borrow
+    /// @param _market The market contract
+    /// @param _spender The `allowanceTarget` field from the API response.
+    /// @param _swapCallData The `data` field from the API response.
+    /// @param _permit Permit data
+    /// @param _helperData Optional helper data in case the collateral needs to be transformed
+    /// @param _dbrData Optional data in case the user wants to buy DBR and also withdraw some DOLA
+    function leveragePosition(
+        uint256 _value,
+        address _market,
+        address _spender,
+        bytes calldata _swapCallData,
+        Permit calldata _permit,
+        bytes calldata _helperData,
+        DBRHelper calldata _dbrData
+    ) public payable nonReentrant dolaSupplyUnchanged {
+        if (address(markets[_market].buySellToken) == address(0))
+            revert MarketNotSet(_market);
+
+        IMarket market = IMarket(_market);
+
+        // Mint and approve
+        _mintAndApproveDola(_spender, _value);
+
+        // Call the encoded swap function call on the contract at `swapTarget`,
+        // passing along any ETH attached to this function call to cover protocol fees.
+        (bool success, ) = exchangeProxy.call{value: msg.value}(_swapCallData);
+        if (!success) revert SwapFailed();
+
+        // Actual collateral/buyToken bought
+        uint256 collateralAmount = markets[_market].buySellToken.balanceOf(address(this));
```

Comment:
Suggestion:
`require(collateralAmount > 0);` ... this protects against potential configuration error where buySellToken doesn't match the output token from the API supplied calldata. In such a scenario, we run the risk of stuck tokens.